pyrealm/demography/community.py

```diff
@@ -0,0 +1,523 @@
+
+        This factory method loads community data from a JSON community file and
+        validates it using
+        `class`:`~pyrealm.demography.community.CommunityStructuredDataSchema` before
+        using the data to initialise a Community instance.
+
+        Args:
+            path: A path to a JSON file of community data
+            flora: A Flora instance providing plant functional types used in the
+                community data
+        """
+
+        # Load the data
+        try:
+            file_data = json.load(open(path))
+        except (FileNotFoundError, json.JSONDecodeError) as excep:
+            raise excep
+
+        # Validate the data
+        try:
+            file_data = CommunityStructuredDataSchema().load(data=file_data)
+        except ValidationError as excep:
+            raise excep
+
+        return cls(**file_data, flora=flora)
+
+    @classmethod
+    def from_toml(cls, path: Path, flora: Flora) -> Community:
+        """Create a Community object from a TOML file.
+
+        This factory method loads community data from a TOML community file and
+        validates it using
+        `class`:`~pyrealm.demography.community.CommunityStructuredDataSchema` before
+        using the data to initialise a Community instance.
+
+        Args:
+            path: A path to a TOML file of community data
+            flora: A Flora instance providing plant functional types used in the
+                community data
+        """
+
+        # Load the data
+        try:
+            file_data = tomllib.load(open(path, "rb"))
+        except (FileNotFoundError, TOMLDecodeError) as excep:
+            raise excep
+
+        # Validate the data
+        try:
+            file_data = CommunityStructuredDataSchema().load(data=file_data)
+        except ValidationError as excep:
+            raise excep
+
+        return cls(**file_data, flora=flora)
+
+    # @classmethod
+    # def load_communities_from_csv(
+    #     cls, cell_area: float, csv_path: str, flora: Flora
+    # ) -> list[Community]:
+    #     """Loads a list of communities from a csv provided in the appropriate format.
+
+    #     The csv should contain the following columns: cell_id,
+    #     diameter_at_breast_height, plant_functional_type, number_of_individuals. Each
+    #     row in the csv should represent one cohort.
+
+    #     :param cell_area: the area of the cell at each location, this is assumed to be
+    #     the same across all the locations in the csv.
+    #     :param csv_path: path to the csv containing community data, as detailed above.
+    #     :param flora: a flora object, ie a dictionary of plant functional properties,
+    #     keyed by pft name.
+    #     :return: a list of community objects, loaded from the csv
+    #     file.
+    #     """
+    #     community_data = pd.read_csv(csv_path)
+
+    #     data_grouped_by_community = community_data.groupby(community_data.cell_id)
+
```

Comment:
Is this surplus to requirements now?

Reply:
No - but I need to think about a common interface to read multiple communities from CSV/TOML/JSON. Could be a class (`CommunityList`?), could be standalone functions. It's there so I don't forget to do that.